File: app/src/main/java/rusha/x/recipe/createOrEdit/CreateOrEditRecipe.java
package rusha.x.recipe.createOrEdit;

import com.google.gson.annotations.SerializedName;

import java.util.List;

public class CreateOrEditRecipe {
    @SerializedName("id")
    public final Integer id;

    @SerializedName("name")
    public final String name;

    @SerializedName("description")
    public final String description;

    @SerializedName("ingredients")
    public final List<CreateOrEditRecipeIngredient> ingredients;

    public CreateOrEditRecipe(Integer id, String name, String description, List<CreateOrEditRecipeIngredient> ingredients) {
        this.id = id;
        this.name = name;
        this.description = description;
        this.ingredients = ingredients;
    }
}

File: app/src/main/java/rusha/x/recipe/createOrEdit/CreateOrEditRecipe.kt
package rusha.x.recipe.createOrEdit

import com.google.gson.annotations.SerializedName


data class CreateOrEditRecipe(
    @SerializedName("id") val id: Int? = null,
    @SerializedName("name") val name: String,
    @SerializedName("description") val description: String,
    @SerializedName("ingredients") val ingredients: List<Ingredient>
) {
    data class Ingredient(
        val countInRecipe: Double,
        val product: CreateOrEditProductByName
    ) {
        fun withIncrementCount(): Ingredient {
            return copy(
                countInRecipe = countInRecipe + 1
            )
        }

        fun withDecrementCount(): Ingredient {
            return copy(
                countInRecipe = (countInRecipe - 1).coerceAtLeast(minimumValue = 0.0)
            )
        }

        fun withCount(count: Double): Ingredient {
            return copy(
                countInRecipe = count
            )
        }
    }
}


data class CreateOrEditProductByName(
    @SerializedName("id") val id: Int? = null,
    @SerializedName("name") val name: String,
    @SerializedName("price") val price: Double,
    @SerializedName("unit") val unit: String
)